### ISSUE TITLE
Combine Control Widgets with Latest Values

### ISSUE THREAD

**Issue by ddomit** [thingsboard/thingsboard#1623]
Hello Everyone,

I'm trying to create a custom widget in which I can get the latest values from a device updating periodically (everytime they change) and from the same widget have a button or some buttons that send RPC commands to the device.

I tried mixing the code from the RPC Commands widget with the Attribute Card widget code, both from the tutorial on how to develop widgets. For some reason I cant send the RPC command, it seems that it needs a "success(response)" and "fail(rejection)", but when I run the RPC code from within the Attribute Card widget, these response and rejection dont seem to be generating. 
 
 
 
 

**### This is what im trying to build:**

![image](https://user-images.githubusercontent.com/6592826/55939766-2e6c8800-5c04-11e9-998e-e081756d4c86.png)

**### This is my Javascript code:**

____________________________________________________________________________________________________________
`self.onInit = function() {
    
    self.ctx.$scope.sendCommand = function() {
            console.log("0");
        // var rpcMethod = self.ctx.$scope.rpcMethod;
        var rpcMethod = "setValue";
        // var rpcParams = self.ctx.$scope.rpcParams;
        var rpcParams = "True";
        var timeout = self.ctx.settings.requestTimeout;
        var oneWayElseTwoWay = self.ctx.settings.oneWayElseTwoWay ? true : false;
        
        var commandPromise;
        if (oneWayElseTwoWay) {
            commandPromise = self.ctx.controlApi.sendOneWayCommand(rpcMethod, rpcParams, timeout);
                console.log("1");
                console.log(timeout)
                console.log(commandPromise);
        } else {
            commandPromise = self.ctx.controlApi.sendTwoWayCommand(rpcMethod, rpcParams, timeout);
                console.log("2");
        }
        
        // console.log("Response:" + response);
        commandPromise.then(
            function success(response) {
                
                if (oneWayElseTwoWay) {
                    self.ctx.$scope.rpcCommandResponse = "Command was successfully received by device.<br/> No response body because of one way command mode.";
                    console.log("3");
                } else {
                    self.ctx.$scope.rpcCommandResponse = "Response from device:<br/>";                    
                    self.ctx.$scope.rpcCommandResponse += angular.toJson(response);
                    console.log("4");
                }
            },
            
            function fail(rejection) {
                console.log("Rejection:" + rejection);
                console.log("5");
                self.ctx.$scope.rpcCommandResponse = "Failed to send command to the device:<br/>"
                console.log("6");
                self.ctx.$scope.rpcCommandResponse += "Status: " + rejection.status + "<br/>";
                self.ctx.$scope.rpcCommandResponse += "Status text: '" + rejection.statusText + "'";
                console.log("7");
            }
            
        );
    }
    
    var button = document.getElementById("THEbutton");
    button.addEventListener("click", changeOwnText);
    
    function changeOwnText(){
        button.innerHTML += "CLICK!";
    }
    
    self.ctx.datasourceTitleCells = [];
    self.ctx.valueCells = [];
    self.ctx.labelCells = [];
    
    for (var i=0; i < self.ctx.datasources.length; i++) {
        var tbDatasource = self.ctx.datasources[i];

        var datasourceId = 'tbDatasource' + i;
        self.ctx.$container.append(
            "<div id='" + datasourceId +
            "' class='tbDatasource-container'></div>"
        );

        var datasourceContainer = $('#' + datasourceId,
            self.ctx.$container);

        datasourceContainer.append(
            "<div class='tbDatasource-title'>" +
            tbDatasource.name + "</div>"
        );
        
        var datasourceTitleCell = $('.tbDatasource-title', datasourceContainer);
        self.ctx.datasourceTitleCells.push(datasourceTitleCell);
        
        var tableId = 'table' + i;
        datasourceContainer.append(
            "<table id='" + tableId +
            "' class='tbDatasource-table'><col width='30%'><col width='70%'></table>"
        );
        var table = $('#' + tableId, self.ctx.$container);

        for (var a = 0; a < tbDatasource.dataKeys.length; a++) {
            var dataKey = tbDatasource.dataKeys[a];
            var labelCellId = 'labelCell' + a;
            var cellId = 'cell' + a;
            table.append("<tr><td id='" + labelCellId + "'>" + dataKey.label +
                "</td><td id='" + cellId +
                "'></td></tr>");
            var labelCell = $('#' + labelCellId, table);
            self.ctx.labelCells.push(labelCell);
            var valueCell = $('#' + cellId, table);
            self.ctx.valueCells.push(valueCell);
        }
    }    
    
    self.onResize();
}

self.onDataUpdated = function() {
    for (var i = 0; i < self.ctx.valueCells.length; i++) {
        var cellData = self.ctx.data[i];
        if (cellData && cellData.data && cellData.data.length > 0) {
            var tvPair = cellData.data[cellData.data.length -
                1];
            var value = tvPair[1];
            var textValue;
            //toDo -> + IsNumber
            
            if (isNumber(value)) {
                var decimals = self.ctx.decimals;
                var units = self.ctx.units;
                if (cellData.dataKey.decimals || cellData.dataKey.decimals === 0) {
                    decimals = cellData.dataKey.decimals;
                }
                if (cellData.dataKey.units) {
                    units = cellData.dataKey.units;
                }
                txtValue = self.ctx.utils.formatValue(value, decimals, units, true);
            } else {
                txtValue = value;
            }
            self.ctx.valueCells[i].html(txtValue);
            document.getElementById("demo").innerHTML = "datasources[0].name: " + self.ctx.datasources[0].name;
            document.getElementById("demo2").innerHTML = "datasources[0].entityName: " + self.ctx.datasources[0].entityName;
            document.getElementById("demo3").innerHTML = "datasources[0].entityType: " + self.ctx.datasources[0].entityType;
            document.getElementById("demo4").innerHTML = "datasources[0].dataKeys: " + String(self.ctx.data[0].data[0]);
            document.getElementById("demo5").innerHTML = "datasources[0].dataKeys: " + String(self.ctx.datasources[0].dataKeys[0].label + ": " + self.ctx.data[0].data[0][1]);
            // console.log("datasources[0].dataKeys: " + self.ctx.utils.formatValue(self.ctx.datasources[0].dataKeys[0].label, decimals, units, true));
        }
    }
    
    function isNumber(n) {
        return !isNaN(parseFloat(n)) && isFinite(n);
    }
}

self.onResize = function() {
    var datasoirceTitleFontSize = self.ctx.height/8;
    if (self.ctx.width/self.ctx.height <= 1.5) {
        datasoirceTitleFontSize = self.ctx.width/12;
    }
    datasoirceTitleFontSize = Math.min(datasoirceTitleFontSize, 20);
    for (var i = 0; i < self.ctx.datasourceTitleCells.length; i++) {
        self.ctx.datasourceTitleCells[i].css('font-size', datasoirceTitleFontSize+'px');
    }
    var valueFontSize = self.ctx.height/9;
    var labelFontSize = self.ctx.height/9;
    if (self.ctx.width/self.ctx.height <= 1.5) {
        valueFontSize = self.ctx.width/15;
        labelFontSize = self.ctx.width/15;
    }
    valueFontSize = Math.min(valueFontSize, 18);
    labelFontSize = Math.min(labelFontSize, 18);

    for (i = 0; i < self.ctx.valueCells; i++) {
        self.ctx.valueCells[i].css('font-size', valueFontSize+'px');
        self.ctx.valueCells[i].css('height', valueFontSize*2.5+'px');
        self.ctx.valueCells[i].css('padding', '0px ' + valueFontSize + 'px');
        self.ctx.labelCells[i].css('font-size', labelFontSize+'px');
        self.ctx.labelCells[i].css('height', labelFontSize*2.5+'px');
        self.ctx.labelCells[i].css('padding', '0px ' + labelFontSize + 'px');
    }    
}

self.onDestroy = function() {
}
`
___________________________________________________________________________________________________________
___________________________________________________________________________________________________________
___________________________________________________________________________________________________________

**### This is what i get when i console.log(commandPromise), and the general error(in red):**

![image](https://user-images.githubusercontent.com/6592826/55941566-10a12200-5c08-11e9-8492-eb9ff8bc014f.png)

**### And this is the error i get when trying to console.log("Rejection" + rejection):**

>Rejection:undefined

_______________________________________________________________________________

I hope someone can guide me in the right direction, I have tried it all but I dont know what Im missing.

**THANKS A LOT IN ADVANCE**

Daniel

**Comment by ddomit** [comment]
**### This is the JSON File of the updated widget   (base code from attribute card, and added JS code from RPC):**


`
{
  "widget": {
    "isSystemType": false,
    "bundleAlias": "html_design",
    "typeAlias": "attributes_card_test",
    "type": "latest",
    "title": "New widget",
    "sizeX": 13,
    "sizeY": 11,
    "config": {
      "datasources": [
        {
          "type": "entity",
          "dataKeys": [
            {
              "name": "altitude",
              "type": "timeseries",
              "label": "altitude",
              "color": "#2196f3",
              "settings": {},
              "_hash": 0.6736835411882687
            },
            {
              "name": "battery_level",
              "type": "timeseries",
              "label": "battery_level",
              "color": "#4caf50",
              "settings": {},
              "_hash": 0.779278766852278
            },
            {
              "name": "battery_voltage",
              "type": "timeseries",
              "label": "battery_voltage",
              "color": "#f44336",
              "settings": {},
              "_hash": 0.22907833087904383
            },
            {
              "name": "enabled",
              "type": "attribute",
              "label": "enabled",
              "color": "#ffc107",
              "settings": {},
              "_hash": 0.30491348534361196
            },
            {
              "name": "active",
              "type": "attribute",
              "label": "active",
              "color": "#607d8b",
              "settings": {},
              "_hash": 0.18955508068889793
            }
          ],
          "entityAliasId": "7eb4eb3b-636c-1580-69a1-b24a920e8e3d"
        }
      ],
      "timewindow": {
        "realtime": {
          "timewindowMs": 60000
        }
      },
      "showTitle": true,
      "backgroundColor": "#fff",
      "color": "rgba(0, 0, 0, 0.87)",
      "padding": "8px",
      "settings": {
        "oneWayElseTwoWay": true,
        "requestTimeout": 500
      },
      "title": "New Attributes card test",
      "dropShadow": true,
      "enableFullscreen": true,
      "widgetStyle": {},
      "titleStyle": {
        "fontSize": "16px",
        "fontWeight": 400
      },
      "useDashboardTimewindow": true,
      "showLegend": false,
      "actions": {}
    },
    "id": "35805fcb-31f0-0887-3251-7f50535543db"
  },
  "aliasesInfo": {
    "datasourceAliases": {
      "0": {
        "alias": "Drone",
        "filter": {
          "type": "singleEntity",
          "resolveMultiple": false,
          "singleEntity": {
            "entityType": "DEVICE",
            "id": "cf202cf0-44e6-11e9-bb4b-532cc653599b"
          }
        }
      }
    },
    "targetDeviceAliases": {}
  },
  "originalSize": {
    "sizeX": 13,
    "sizeY": 11
  },
  "originalColumns": 24
}
`

**Comment by bhumika** [comment]
Hello Ddomit,

I tried to import the provided sample JSON file.  Getting below error:
**"Unable to import widgets bundle: Invalid widgets bundle data structure."**

Please help.

-Bhumika

**Comment by lvojp** [comment]
Hello everyone.
I'm just worried about this issue right now.

Specifically, I want to press the button, wait for the device to respond, and display the data returned from the device in the widget.

Is there a way to solve this specification?

Thank you for your teaching.

**Comment by marianodo** [comment]
Hi @Ivojp 
Could combine control and last value widget?
I need the same kind of combination
Thanks
Mariano

**Comment by IvoJP** [comment]
@marianodo I think you tagged the wrong person. I think you're looking for @lvojp (L instead of I - an unfortunate font choice by github. Perhaps @lvojp can capitalize that L ;) )

You might want to add another comment and tag the appropriate user.

**Comment by marianodo** [comment]
@IvoJP yes, sorry, my mistake. 
I edited the comment.
Thanks

**Comment by momolok** [comment]
@ddomit, did you manage to get this work? I'm currently facing the same challenge of receiving telemetry and sending RPC commands, which is partly conditioned by received the telemetry data, using the same widget.

**Comment by paolofrs** [comment]
Hi, any news here? I know there's should be a way to do so (RPC/Control Widget + Last Value) since the default Switch Control widget (in the Control Widget bundle) there is the possibility to update the status of the switch with a specific telemetry instead of the Response from SendTwoWayCommand. I wonder how to do it in custom widget...

**Comment by PeterMair** [comment]
Hi @paolofrs,


it isn't possible to use **self.ctx.controlApi.sendTwoWayCommand** on a time/last value widget.
I use this solution for **sendOneWayCommand**, but you can replace **sendOneWayRpcCommand** with **sendTwoWayRpcCommand**:

```
// Get deviceid from first datasource

var deviceid = self.ctx.datasources[0].entityId;

// Get deviceService 

var $injector = self.ctx.$scope.$injector;
var $deviceservice = $injector.get(self.ctx.servicesMap.get('deviceService'));

// Send oneWay RPC-Command, with requestBody like {"method": "methodname", "params": 1} 
      
commandPromise  = $deviceservice.sendOneWayRpcCommand(deviceid, requestBody, { ignoreErrors: true  });
commandPromise.subscribe(
  function success(response) {
   console.log("Command was successfully received by device. No response body because of one way command mode.");       
  },
  
  function fail(rejection) {
      
      console.log("Failed to send command to the device:");
      console.log(rejection);
      
  }
);
```

**Comment by paolofrs** [comment]
Hi @PeterMair 

Thank you very much! From #4085 seems that someone succeed with your solution! Anyway, I am a bit confused. 

I've created a **latest value** widget and added [your piece of code](https://github.com/thingsboard/thingsboard/issues/1623#issuecomment-743993310) following the [RPC widget development official guide](https://thingsboard.io/docs/user-guide/contribution/widgets-development/#rpc-control-widget) but if does not work. Here it is:

~~~
self.onInit = function() {
    
    self.ctx.$scope.sendCommand = function() {
        var rpcMethod = self.ctx.settings.rpcMethod;
        var rpcParams = self.ctx.settings.rpcParams;
        var timeOutSet = self.ctx.$scope.timeOutSet;
        var oneWayElseTwoWay =  true; // always TRUE with LoRa
        var timeout = 500; // set to 500 ms fixed

        // Get deviceid from first datasource
        var deviceid = self.ctx.datasources[0].entityId;
    
        // Get deviceService 
        var $injector = self.ctx.$scope.$injector;
        var $deviceservice = $injector.get(self.ctx.servicesMap.get('deviceService'));

        // Send oneWay RPC-Command, with requestBody like {"method": "methodname", "params": 1} 
        commandPromise  = $deviceservice.sendOneWayRpcCommand(deviceid, requestBody, { ignoreErrors: true  });
        
        commandPromise.subscribe(
            function success(response) {
                console.log("Command was successfully received by device. No response body because of one way command mode.");       
            },
            function fail(rejection) {
                console.log("Failed to send command to the device:");
                console.log(rejection);
            }
        );
    }    
}
~~~

The widget does not even render in the bottom right corner... If I switch to **Control Widget** widget type it renders but it returns this error when pressing the md-button associated with  `sendCommand()` function (from browser debug console):

~~~
ERROR TypeError: self.ctx.datasources is undefined
~~~


What am I doing wrong? Can you provide more details?

**Comment by paolofrs** [comment]
Fixed with #4085, [hjlopezr](https://github.com/hjlopezr) helped me to find my error! 

I wrote a small explanation if anyonelse lands here... The solution for me was to add the `sendOneWayRpcCommand` function to a `lastvalue` widget.

**Comment by Vaibhav0216** [comment]
`commandPromise is not defined
    at ShowDataComponent.rpc_request `  this error comes at below code what should I need to update. 

```
self.onInit = function() {
    const showDataComponent = new ShowDataComponent();

   // Select buttons with the class `.control-button`
   const buttons = document.querySelectorAll('.toggle-btn');
   if (buttons.length === 0) {
       console.error("No buttons found with class 'toggle-btn'.");
   }

   buttons.forEach(button => {
       button.addEventListener('click', (event) => {
           showDataComponent.toggleState(event.target);
       });
   });

   console.log("Event listeners attached to buttons:", buttons);
   self.ctx.datasourceTitleCells = [];
   self.ctx.valueCells = [];
   self.ctx.labelCells = [];
}

class ShowDataComponent {
   constructor() {
   }
   
  toggleState(button) {
       // Debug: Log button details
       console.log("Button clicked:", button);
   
       // Remember the initial state
       const wasEnabled = button.classList.contains("on");
   
       // Toggle the button's state optimistically
       const isEnabled = button.classList.toggle("on");
       button.classList.toggle("off", !isEnabled);
   
       // Update the button text
       button.textContent = isEnabled ? "Enable" : "Disable";
   
       const pumpName = button.id;
       let cid = -1;
       const state = isEnabled ? 1 : 0;
       if (pumpName === "pump_1") {
           cid = 82; 
       }
       } else if(pumpName === "pump_2"){
           cid = 83;
       } else {
           console("not clicked on the button");
           return;
       }
       // Debug: Log state changes
       console.log(`${pumpName} state updated to: ${state}`);
   
       // Payload to send
      const payload = JSON.stringify({
          method : "COIL",
          params: {
              cid: cid,
              state: state
          }
      });
      this.rpc_request(payload);
   }
   rpc_request(payload) {        var timeOutSet = self.ctx.$scope.timeOutSet;
        var oneWayElseTwoWay =  true; // always TRUE with LoRa
        var timeout = 500; // set to 500 ms fixed

        // Get deviceid from first datasource
        var deviceid = self.ctx.datasources[0].entityId;
    
        // Get deviceService 
        var $injector = self.ctx.$scope.$injector;
        var $deviceservice = $injector.get(self.ctx.servicesMap.get('deviceService'));

        // Send oneWay RPC-Command, with requestBody like {"method": "methodname", "params": 1} 
        commandPromise  = $deviceservice.sendOneWayRpcCommand(deviceid, payload, { ignoreErrors: true  });
        
        console.log("test seuccess");
        
        commandPromise.subscribe(
            function success(response) {
                console.log("Command was successfully received by device. No response body because of one way command mode.");       
            },
            function fail(rejection) {
                console.log("Failed to send command to the device:");
                console.log(rejection);
            }
        );
    }    

}

self.onDataUpdated = function() {
   // Call getEntityTimeseriesLatest and process the data
   const entityId = { entityType: 'DEVICE', id: '119f0c50-b218-11ef-b008-21b9a1a1df1b' }; // Example device ID
   const keys= [
       "Pump_1",
       "Pump_2",
     ];

  // Make the API call to fetch the latest timeseries data
   self.ctx.attributeService.getEntityTimeseriesLatest(entityId, keys).subscribe(
       (data) => {
           // Log the data for debugging
           console.log(data);

           // Process and update the data in the HTML elements
           updateData(data); // Call the updateData function with the fetched data
       },
       (error) => {
           console.error('Error fetching telemetry data:', error);
       }
   );
    // The updateData function to update the HTML elements
   function  updateData(data) {
     console.log(data);
      for (const [key, values] of Object.entries(data)) {
       const buttonId = key.toLowerCase(); // Convert to lowercase to match button IDs
       const value = values[0]?.value === "true"; // Extract the value field and convert to boolean 
       console.log("buttonId", buttonId);
       console.log("value",value);
       // Find the button element
       const button = document.getElementById(buttonId);
       if (button) {
         // Update button state based on value
         button.classList.toggle("on", value);
         button.classList.toggle("off", !value);
         button.textContent = value ? "Enable" : "Disable";
       }
     }
 }

};
```
